modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheMetricsImpl.java

```diff
@@ -844,6 +859,43 @@ public void onRead(boolean isHit) {
             delegate.onRead(isHit);
     }
 
+    /** Set callback for tx key collisions detection. */
+    public void keyCollisionsInfo(Supplier<List<Map.Entry<GridCacheMapEntry, Integer>>> coll) {
+        txKeyCollisionInfo = coll;
+
+        if (delegate != null)
+            delegate.keyCollisionsInfo(coll);
+    }
+
+    /** Get existing callback. */
+    public @Nullable Supplier<List<Map.Entry<GridCacheMapEntry, Integer>>> keyCollisionsInfo() {
+        return txKeyCollisionInfo;
+    }
+
+    /** {@inheritDoc} */
+    @Override public String getTxKeyCollisions() {
+        SB sb = null;
+
+        if (txKeyCollisionInfo != null) {
+            List<Map.Entry<GridCacheMapEntry, Integer>> result = txKeyCollisionInfo.get();
```

Comment:
There is no guarantee that you will not get an npe, let's start by putting a supplier in local variable.